api/src/main/java/org/apache/iceberg/actions/compaction/DataCompactionStrategy.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions.compaction;
+
+import java.io.Serializable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.expressions.Expression;
+
+public interface DataCompactionStrategy extends Serializable {
+  /**
+   * Returns the name of this compaction strategy
+   */
+  String name();
+
+  /**
+   * Returns a set of options which this compaction strategy can use. This is an allowed-list and any options not
+   * specified here will be rejected at runtime.
+   */
+  Set<String> validOptions();
+
+  DataCompactionStrategy withOptions(Map<String, String> options);
+
+  /**
+   * Before the compaction strategy rules are applied, the underlying action has the ability to use this expression to
+   * filter the FileScanTasks which will be created when planning file reads that will later be run through this
+   * compaction strategy. This would be used for pushing down filter expressions to the manifest entry scanning phase.
+   *
+   * @return an Iceberg expression to use when discovering file scan tasks
+   */
+  Expression preFilter();
+
+  /**
+   * Removes all file references which this plan will not rewrite or change. Unlike the preFilter, this method can
+   * execute arbitrary code and isn't restricted to just Iceberg Expressions. This should be serializable so that
+   * Actions which run remotely can utilize the method.
+   *
+   * @param dataFiles iterator of live datafiles in a given partition
+   * @return iterator containing only files to be rewritten
+   */
+  Iterator<FileScanTask> filesToCompact(Iterator<FileScanTask> dataFiles);
```

Comment:
Do we think working with `Iterator` would be easier than with `Iterable`? Just out of curiosity.

Reply:
I don't have a strong feeling on this, I initially wanted to stress the single pass nature to try to allow frameworks to pass lazy iterators (like spark) but i'm beginning to think all of our actual implementations for more complicated things will need to materialize the whole set of tasks anyway ...

Reply:
I generally prefer `Iterable`, but is it a good idea to work on a collection here? Why not `canSkip(FileScanTask)` or something?

Reply:
With Sort for example we need to see a file in context with others to know if we are going to keep it. So either we have access to the whole set of files here, or the implementer needs to keep static state about previous file's seen.

Reply:
I think we mostly use `Iterable` in core and Guava transforms that are lazy. It would be nice to express this as a filter but the proposed strategies will need to know a broader picture so I think it should be fine to operate on iterables or iterator here.

Reply:
This is used to select candidates from a set of files, how about calling it with select or filter prefix?